examples/ws_guessing_game.rs

```diff
@@ -0,0 +1,236 @@
+            warn!("Not a number: `{}` (length {})", input, input.len());
+            return None;
+        };
+        if !(1..=100).contains(&number) {
+            warn!("Not in range ({})", number);
+            return None;
+        }
+        Some(number)
+    }
+}
+
+// Super rudimentary pseudo-random numbers
+fn rand() -> u32 {
+    EspSystemTime::now(&EspSystemTime {}).subsec_nanos() / 65537
+}
+
+// Serialize numbers in English
+fn nth(n: u32) -> Cow<'static, str> {
+    match n {
+        smaller @ (0..=13) => Cow::Borrowed(match smaller {
+            0 => "zeroth",
+            1 => "first",
+            2 => "second",
+            3 => "third",
+            4 => "fourth",
+            5 => "fifth",
+            6 => "sixth",
+            7 => "seventh",
+            8 => "eighth",
+            9 => "ninth",
+            10 => "10th",
+            11 => "11th",
+            12 => "12th",
+            13 => "13th",
+            _ => unreachable!(),
+        }),
+        larger => Cow::Owned(match larger % 10 {
+            1 => format!("{}st", larger),
+            2 => format!("{}nd", larger),
+            3 => format!("{}rd", larger),
+            _ => format!("{}th", larger),
+        }),
+    }
+}
+
+fn main() -> anyhow::Result<()> {
+    esp_idf_sys::link_patches();
+    esp_idf_svc::log::EspLogger::initialize_default();
+    let mut server = create_server()?;
+
+    server.fn_handler("/", Method::Get, |req| {
+        req.into_ok_response()?.write(INDEX_HTML.as_bytes())?;
+        Ok(())
+    })?;
+
+    let guessing_games = Mutex::new(BTreeMap::<i32, GuessingGame>::new());
+
+    server
+        .ws_handler("/ws/guess", move |ws| {
+            let mut sessions = guessing_games.lock().unwrap();
+            if ws.is_new() {
+                sessions.insert(ws.session(), GuessingGame::new((rand() % 100) + 1));
+                info!("New WebSocket session ({} open)", sessions.len());
+                ws.send(
+                    FrameType::Text(false),
+                    "Welcome to the guessing game! Enter a number between 1 and 100".as_bytes(),
+                )?;
+                return Ok(());
+            } else if ws.is_closed() {
+                sessions.remove(&ws.session());
+                info!("Closed WebSocket session ({} open)", sessions.len());
+                return Ok(());
+            }
+            let session = sessions.get_mut(&ws.session()).unwrap();
+
+            // NOTE: Due to the way the underlying C implementation works, ws.recv()
+            // may only be called with an empty buffer exactly once to receive the
+            // incoming buffer size, then must be called exactly once to receive the
+            // actual payload.
+
+            let (_frame_type, len) = match ws.recv(&mut []) {
+                Ok(frame) => frame,
+                Err(e) => return Err(e),
+            };
+
+            if len > MAX_LEN {
```

Comment:
with this specific MAX_LEN can  you even trigger a Request too big? A frame should be bounded by the underlying packet MTU size or am i wrong here, in which case we would get spitted packages, and would never come close to 10KB

Reply:
WS frames run on top of TCP/IP so MTU does not apply to them.

Reply:
You can return any `esp_error_t` from any handler and the socket will be closed as per the IDF documentation. The specific error type used is pretty much up to the socket handler. ESP_FAIL is good but it's better to be more descriptive.